File: packages/election_rules_py/election_rules_py/candidate_probability.py
"""Candidate seat probability, coalition majority probability, and
simulation summary statistics.

Direct port of packages/election_rules/lib/src/candidate_probability.dart —
keep in sync.
"""

from __future__ import annotations

import bisect
from dataclasses import dataclass


def candidate_seat_probability(rank: int, party_seat_simulations: list[int]) -> float:
    """Probability a closed-list candidate at `rank` (1-based) wins a seat.

    Under closed-list PR, a voter votes for the list, not the candidate.
    A candidate at `rank` wins a seat in a given simulation iff that
    list's simulated seat count in that iteration is >= rank. This is
    CRITICAL ACCURACY RULE #7's implementation: never derive this from a
    fabricated per-candidate vote share.
    """
    if rank < 1:
        raise ValueError("Candidate list rank must be >= 1")
    if not party_seat_simulations:
        return 0.0
    win_count = sum(1 for seats in party_seat_simulations if seats >= rank)
    return win_count / len(party_seat_simulations)


def candidate_seat_probabilities_for_list(
    list_size: int, party_seat_simulations: list[int]
) -> list[float]:
    """Probability per rank (index 0 = rank 1) for an entire list in one pass.

    Guarantees monotonicity: probability(rank) never increases as rank
    increases, satisfying the acceptance test that candidate #10 cannot
    have a lower probability than #11 on the same normal closed list.
    """
    if not party_seat_simulations:
        return [0.0] * list_size
    n = len(party_seat_simulations)
    sorted_sims = sorted(party_seat_simulations)
    result = []
    for i in range(list_size):
        rank = i + 1
        lo = bisect.bisect_left(sorted_sims, rank)
        count_at_least_rank = len(sorted_sims) - lo
        result.append(count_at_least_rank / n)
    return result


def majority_threshold(total_seats: int) -> int:
    """floor(total_seats / 2) + 1 — never hard-coded elsewhere."""
    return (total_seats // 2) + 1


def coalition_majority_probability(
    coalition_list_ids: set[str],
    simulations: list[dict[str, int]],
    total_seats: int,
) -> float:
    """Probability that the combined seats of coalition_list_ids meet or
    exceed the majority threshold, across Monte Carlo simulations.
    """
    if not simulations:
        return 0.0
    threshold = majority_threshold(total_seats)
    hits = 0
    for sim in simulations:
        combined = sum(sim.get(list_id, 0) for list_id in coalition_list_ids)
        if combined >= threshold:
            hits += 1
    return hits / len(simulations)


@dataclass(frozen=True)
class SimulationSummary:
    median: float
    mean: float
    low50: float
    high50: float
    low80: float
    high80: float
    low95: float
    high95: float

    @staticmethod
    def from_values(values: list[float]) -> "SimulationSummary":
        if not values:
            return SimulationSummary(0, 0, 0, 0, 0, 0, 0, 0)
        sorted_values = sorted(values)
        n = len(sorted_values)

        def pct(p: float) -> float:
            idx = min(max(round((n - 1) * p), 0), n - 1)
            return float(sorted_values[idx])

        mean = sum(sorted_values) / n
        return SimulationSummary(
            median=pct(0.5),
            mean=mean,
            low50=pct(0.25),
            high50=pct(0.75),
            low80=pct(0.10),
            high80=pct(0.90),
            low95=pct(0.025),
            high95=pct(0.975),
        )

File: packages/election_rules/lib/src/candidate_probability.dart
import 'rule_set.dart';

/// Computes a closed-list candidate's estimated probability of entering
/// parliament from a set of Monte Carlo simulation outcomes.
///
/// Under closed-list PR, a voter votes for the list, not the candidate.
/// A candidate at [rank] (1-based position on the list) wins a seat in a
/// given simulation iff that list's simulated seat count in that
/// iteration is >= [rank]. This is CRITICAL ACCURACY RULE #7's
/// implementation: we never derive this from a fabricated per-candidate
/// vote share.
///
/// [partySeatSimulations] is one simulated seat count per Monte Carlo
/// iteration for the candidate's list (same list, across all iterations
/// of one forecast run).
///
/// Returns a probability in [0, 1]. Callers are responsible for rounding
/// for display (never show more than whole-percent precision — see
/// CRITICAL ACCURACY RULE #9 / section 21).
double candidateSeatProbability({
  required int rank,
  required List<int> partySeatSimulations,
}) {
  if (rank < 1) {
    throw ArgumentError.value(rank, 'rank', 'Candidate list rank must be >= 1');
  }
  if (partySeatSimulations.isEmpty) return 0;

  final winCount = partySeatSimulations.where((seats) => seats >= rank).length;
  return winCount / partySeatSimulations.length;
}

/// Computes candidateSeatProbability for every rank on a list in one pass,
/// guaranteeing the monotonicity property required by the acceptance
/// tests: probability(rank) must never increase as rank increases.
List<double> candidateSeatProbabilitiesForList({
  required int listSize,
  required List<int> partySeatSimulations,
}) {
  if (partySeatSimulations.isEmpty) return List.filled(listSize, 0);
  final n = partySeatSimulations.length;
  // Sort descending once; probability(rank) = count(seats >= rank) / n,
  // which is monotonically non-increasing in rank by construction.
  final sorted = [...partySeatSimulations]..sort();
  return List.generate(listSize, (i) {
    final rank = i + 1;
    // Binary search for first index >= rank in ascending sorted list.
    var lo = 0, hi = sorted.length;
    while (lo < hi) {
      final mid = (lo + hi) >> 1;
      if (sorted[mid] < rank) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }
    final countAtLeastRank = sorted.length - lo;
    return countAtLeastRank / n;
  });
}

/// Majority threshold for a parliament of [totalSeats], never hard-coded
/// as a literal elsewhere in the codebase: floor(totalSeats/2) + 1.
int majorityThreshold(int totalSeats) => (totalSeats ~/ 2) + 1;

/// Probability that the combined seats of [coalitionListIds] meet or
/// exceed the majority threshold, across a set of Monte Carlo simulations.
///
/// [simulations] is a list of per-iteration seat maps (listId -> seats),
/// one map per Monte Carlo iteration, all produced by the same forecast
/// run (see forecasting/monte_carlo.py on the backend for the generator;
/// this function only consumes results, it does not simulate).
double coalitionMajorityProbability({
  required Set<String> coalitionListIds,
  required List<Map<String, int>> simulations,
  required int totalSeats,
}) {
  if (simulations.isEmpty) return 0;
  final threshold = majorityThreshold(totalSeats);
  var hits = 0;
  for (final sim in simulations) {
    final combined = coalitionListIds.fold<int>(0, (sum, id) => sum + (sim[id] ?? 0));
    if (combined >= threshold) hits++;
  }
  return hits / simulations.length;
}

/// Distribution summary (median / mean / credible interval) for a metric
/// (typically seats or vote share) across Monte Carlo iterations.
class SimulationSummary {
  final double median;
  final double mean;
  final double low50;
  final double high50;
  final double low80;
  final double high80;
  final double low95;
  final double high95;

  const SimulationSummary({
    required this.median,
    required this.mean,
    required this.low50,
    required this.high50,
    required this.low80,
    required this.high80,
    required this.low95,
    required this.high95,
  });

  factory SimulationSummary.fromValues(List<num> values) {
    if (values.isEmpty) {
      return const SimulationSummary(
        median: 0, mean: 0, low50: 0, high50: 0, low80: 0, high80: 0, low95: 0, high95: 0,
      );
    }
    final sorted = [...values]..sort();
    double pct(double p) {
      final idx = ((sorted.length - 1) * p).round().clamp(0, sorted.length - 1);
      return sorted[idx].toDouble();
    }

    final mean = sorted.fold<double>(0, (a, b) => a + b) / sorted.length;
    return SimulationSummary(
      median: pct(0.5),
      mean: mean,
      low50: pct(0.25),
      high50: pct(0.75),
      low80: pct(0.10),
      high80: pct(0.90),
      low95: pct(0.025),
      high95: pct(0.975),
    );
  }
}

/// Named reference to the rule set (re-exported here for convenience so
/// forecast-facing code only needs one import).
typedef ElectionRules = ElectionRuleSet;
